### PR TITLE
Upload artifacts to S3 bucket

### PR DESCRIPTION
**Pull request by ntampakas**
**Description**

With this PR a new workflow is introduced that uploads and updates the artifacts to an S3 bucket.
Additionally, a CloudFront distribution for this specific bucket was created to accelerate content delivery.
The distribution domain name is **snark-artifacts.pse.dev**, and files can be downloaded using the following structure:
https://snark-artifacts.pse.dev/poseidon/poseidon-1.wasm
https://snark-artifacts.pse.dev/semaphore-identity/semaphore-identity.wasm

**Issue**

Re #89

### CODE REVIEW

File: .github/scripts/upload-artifacts.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+s3_bucket="snark-artifacts"
+artifacts=$(ls packages/ | grep -v 'artifacts\|cli')
+
+cd packages
+git restore-mtime
+
+for artifact in $artifacts; do
+  aws s3 sync $artifact s3://$s3_bucket/$artifact --delete
+done
```
Comment (sripwoud) [inline .github/scripts/upload-artifacts.sh:12]: as discussed on discord, this is great to deal with a single file version.

snark artifacts is about packaging the artifacts as npm packages, and then distributing them via CDN to allow consumers download them at runtime (and not bloat their deps).

So a requirement is that the bucket must serve all the versions that exists in the npm registry.

could we have something like
`/project/version/file`

e.g

`/poseidon/1.0.0/poseidon-1.zkey`
...

This is what I tried to implement in 
https://github.com/privacy-scaling-explorations/snark-artifacts/pull/90

where I read the package version from the package.json before copying to the bucket

**Review by sripwoud — CHANGES_REQUESTED** [review]

File: .github/scripts/upload-artifacts.sh (hunk 1 above)
Comment (ntampakas) [reply .github/scripts/upload-artifacts.sh:12]: > as discussed on discord, this is great to deal with a single file version.
> 
> snark artifacts is about packaging the artifacts as npm packages, and then distributing them via CDN to allow consumers download them at runtime (and not bloat their deps).
> 
> So a requirement is that the bucket must serve **all** the versions that exist in the npm registry.
> 
> could we have something like `/project/version/file`
> 
> e.g
> 
> `/poseidon/1.0.0/poseidon-1.zkey` `/semaphore/4.0.0-beta.9/semaphore-2.wasm` etc ...
> 
> This is what I tried to implement in #90
> 
> where I read the package version from the package.json before copying to the bucket

Sounds good!
Applying a change in the aux script to upload the specific files to their respective versions.

**Review by sripwoud — APPROVED** [review]
utACK

@ntampakas Should we merge now?

@cedoor maybe the version update for the poseidon package (beta instead of stable) can serve as a test for this workflow.

File: .github/workflows/upload-artifacts.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+name: Upload Artifacts
+on:
+  push:
+    branches:
+      - main
+    paths:
+      - 'packages/**'
```
Comment (sripwoud) [inline .github/workflows/upload-artifacts.yaml:7]: what about?
```suggestion
      - 'packages/**'
      - '!packages/artifacts/**'
      - '!packages/cli/**'
```

Comment (cedoor) [reply .github/workflows/upload-artifacts.yaml:7]: +1

**Comment by ntampakas** [comment]
> utACK
> 
> @ntampakas Should we merge now?
> 
> @cedoor maybe the version update for the poseidon package (beta instead of stable) can serve as a test for this workflow.

I think we are good to go!

**Review by cedoor — APPROVED** [review]

**Pull request merged**